### PR TITLE
Add a simple variant implementation similiar to std::variant

### PR DESCRIPTION
**Pull request by kghost**

### CODE REVIEW

File: src/lib/support/tests/TestVariant.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Unit tests for the Chip Variant API.
```
Comment (andy31415) [inline src/lib/support/tests/TestVariant.cpp:21]: delete this comment block

**Review by andy31415 — APPROVED** [review]
Where do we expect this to be typically used?

One concern for templates is always code size impact. What is the impact of this compared to a more typical (and less safe) union of structures?

Could you do some analysis on how much we would be paying for having more typesafety/convenience here?

**Comment by gerickson** [comment]
> Where do we expect this to be typically used?
> 
> One concern for templates is always code size impact. What is the impact of this compared to a more typical (and less safe) union of structures?
> 
> Could you do some analysis on how much we would be paying for having more typesafety/convenience here?

+1. Took a look at std::variant and this seems to add a substantial amount of reader complexity to the code, never mind what the code size implications might be as @andreilitvin cites.

**Comment by andy31415** [comment]
I believe we proved the overhead is light and we get some type safety, so I would be ok with the PR.

I also believe that usages should be infrequent: it is harder to reason about anything that can be 'X or Y' in most cases, so wherever we use the variant type we should consider if it is really justified (e.g. some sort of generic event queue may be) or just makes implementations 2x or more complicated by having implementations be one large if for all branches.

Nothing against the PR itself or usage, may have design complaints if the variant type gets used too much.

**Review by msandstedt — APPROVED** [review]

**Comment by woody-apple** [comment]
Approving given @andy31415 is OK with this... personally a little worried about this type of thing proliferating heavily.

**Review by woody-apple — APPROVED** [review]

**Pull request merged**